lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp

```diff
@@ -0,0 +1,270 @@
+  void setLastConnect(Value dest, Operation *connection) {
+    // Try to insert, if it doesn't insert, replace the previous value.
+    auto itAndInserted = scope.insert({dest, connection});
+    if (!std::get<1>(itAndInserted)) {
+      auto iterator = std::get<0>(itAndInserted);
+      iterator->second->erase();
+      iterator->second = connection;
+    }
+  }
+
+  /// Get the source value from a connection. This supports any operation which
+  /// is capable of driving a value.
+  static Value getConnectedValue(Operation *op) {
+    return TypeSwitch<Operation *, Value>(op)
+        .Case<ConnectOp>([](auto op) { return op.src(); })
+        .Case<PartialConnectOp>([](auto op) { return op.src(); })
+        .Default([](Operation *op) {
+          llvm_unreachable("unknown operation");
+          return nullptr;
+        });
+  }
+
+  /// And a 1-bit value with the current condition.  If we are in the outer
+  /// scope, i.e. not in a WhenOp region, then there is no condition.
+  Value andWithCondition(Operation *op, Value value) {
+    // If there is no condition, it means we are not inside any WhenOp.  This
+    // means we don't have to combine the value with anything.
+    if (!condition)
+      return value;
+    // 'and' the value with the current condition.
+    return OpBuilder(op).createOrFold<AndPrimOp>(
+        condition.getLoc(), condition.getType(), condition, value);
+  }
+
+  /// Take two connection operations and merge them in to a new connect under a
+  /// condition.  Destination of both connects should be `dest`.
+  static Operation *flattenConditionalConnections(OpBuilder &b, Location loc,
+                                                  Value dest, Value cond,
+                                                  Operation *whenTrueConn,
+                                                  Operation *whenFalseConn) {
+    auto whenTrue = getConnectedValue(whenTrueConn);
+    auto whenFalse = getConnectedValue(whenFalseConn);
+    auto newValue = b.createOrFold<MuxPrimOp>(loc, whenTrue.getType(), cond,
+                                              whenTrue, whenFalse);
+    auto newConnect = b.create<ConnectOp>(loc, dest, newValue);
+    whenTrueConn->erase();
+    whenFalseConn->erase();
+    return newConnect;
+  }
+
+private:
+  /// Map of destinations and the operation which is driving a value to it in
+  /// the current scope. This is used for resolving last connect semantics, and
+  /// for retreiving the resposible connect operation.
+  DenseMap<Value, Operation *> &scope;
+
+  /// The current wrapping condition. If null, we are in the outer scope.
+  Value condition;
+};
+} // namespace
+
+void ExpandWhensVisitor::run(Block &block) {
+  for (auto &op : llvm::make_early_inc_range(block)) {
+    dispatchVisitor(&op);
+  }
+}
+
+void ExpandWhensVisitor::visitStmt(ConnectOp op) {
+  setLastConnect(op.dest(), op);
+}
+
+void ExpandWhensVisitor::visitStmt(PartialConnectOp op) {
+  setLastConnect(op.dest(), op);
```

Comment:
I'm not sure if they are legal here or not.  PartialConnects are not yet handled by LowerTypes, and they could be lowered completely to ConnectOps.

Reply:
LowerTypes is not going to lower partial connects to regular connects, we are going to rely on later canonicalization to handle partial connect -> connect+padding. I think I'm going to keep partial connect support here, so the pass is more flexible in where it can go in the pipeline.